collect rbd image and snap info per rados namespace

yati1998 opened the pull request:
This commit collects rbd image and snap info for each rados namespace

Madhu-1 commented on collection-scripts/gather_ceph_resources:
```diff
@@ -257,7 +257,6 @@ for ns in $namespaces; do
                     printf "Collecting image status for: %s/%s\n" "${bp}" "${image}"
                     timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd status $image --pool $bp" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-status-"${image}"-debug.log
                     printf "Collecting snap info for: %s/%s\n" "${bp}" "${image}"
-                    timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd snap ls --all $image --pool $bp" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-snap-"${image}"-debug.log
```
why we have removed this one?

yati1998 replied:
in next line we are collecting same details with proper format, so it's just as duplication.

jarrpa replied:
But why was it there in the first place?

yati1998 replied:
not sure, what added it.

jarrpa replied:
It looks like there was actually a reason for adding the `--format=json` version: https://github.com/red-hat-storage/ocs-operator/commit/08d7f78e1db1c0414526a27288313be8ffb6be60 Please either return the line you removed or  confirm with Ilya that it's effectively redundant. I personally don't like to try and parse JSON if I don't need to.

jarrpa replied:
Similarly, if you add it back in be sure to add it for the rns block below as well.

yati1998 replied:
if we have the json formatted output, we don't need this. IMO, we can remove this.

Madhu-1 commented on collection-scripts/gather_ceph_resources:
```diff
@@ -269,7 +268,34 @@ for ns in $namespaces; do
-        done
+
+            rados_namespaces=$(timeout 60 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd -p $bp namespace ls")
+            for rns in $rados_namespaces; do
```
add radosnamespace to below code as well

```
ceph_resources=()
ceph_resources+=(cephblockpools)
ceph_resources+=(cephfilesystems)
ceph_resources+=(cephclient)
ceph_resources+=(cephfilesystemsubvolumegroups)
```

jarrpa replied:
+1

yati1998 replied:
`oc adm inspect --dest-dir="${CEPH_COLLECTION_PATH}" --all-namespaces "${resource}" ` we already collect all namespace for these resources, I don't think we need to add anything seperately.

jarrpa replied:
...correct me if I'm wrong, but the `--all-namespaces` here refers to k8s namespaces, and not RADOS namespaces. We're asking you to add `ceph_resources+=(cephblockpoolradosnamespaces)` to gather the rns CRs.

yati1998 replied:
oh is it, sorry, I will check and update accordingly.

Madhu-1 reviewed (changes requested):
* Need to adjust omap code to collect omap in each radosnamespace for both pvc and snapshot
* trash ls need to be collected for radosnamespace

Okay, we need to collect all the details we are collecting for each rbd pool. we need to like below

* Move all the collection code to helper functions
* Collect all details for the pool and also collect all details for each namespace inside the pool.

jarrpa commented on collection-scripts/gather_ceph_resources:
```diff
@@ -269,7 +268,34 @@ for ns in $namespaces; do
+                for image in $images; do
+                    dbglogf "${CEPH_GATHER_DBGLOG}" "collecting vol and snapshot info for ${image}"
+                    {
+                        printf "Collecting image info for: %s/%s\n" "${bp}" "${image}"
```
This and the other printf statements like it need `"${rns}"` added.

yati1998 commented:
@Madhu-1 @jarrpa please do have a look, if these details are all we need.

Madhu-1 commented on collection-scripts/gather_ceph_resources:
```diff
@@ -269,7 +261,64 @@ for ns in $namespaces; do
+
+            rados_namespaces=$(timeout 60 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd -p $bp namespace ls")
+            for rns in $rados_namespaces; do 
+                COMMAND_OUTPUT_FILE=${COMMAND_OUTPUT_DIR}/rbd_vol_and_snap_info_"${bp}"_"${rns}"               
+                dbglogf "${CEPH_GATHER_DBGLOG}" "Collecting image and snap info for images in: %s %s\n" "${bp}" "${rns}"
+                images=$(timeout 60 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd ls -p $bp --namespace $rns")
+                pids_rbd=()
+                for image in $images; do
+                    dbglogf "${CEPH_GATHER_DBGLOG}" "collecting vol and snapshot info for ${image}"
+                    {
+                        printf "Collecting image info for: %s/%s%s\n" "${bp}" "${image}" "${rns}"
+                        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd info $image --pool $bp --namespace $rns" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-"${image}"-debug.log
+                        printf "Collecting image status for: %s/%s%s\n" "${bp}" "${image}" "${rns}"
+                        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd status $image --pool $bp --namespace $rns" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-status-"${image}"-debug.log
+                        printf "Collecting snap info for: %s/%s%s\n" "${bp}" "${image}" "${rns}"
+                        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd snap ls --all --format=json --pretty-format $image --pool $bp --namespace $rns" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-snap-json-"${image}"-debug.log
+                    } >>"${COMMAND_OUTPUT_DIR}"/rbd_vol_and_snap_info_"${image}".part &
+                    pids_rbd+=($!)
+                done
+                if [ -n "${pids_rbd[*]}" ]; then
+                    # wait for all pids
+                    dbglog "waiting for ${pids_rbd[*]} to terminate"
+                    wait "${pids_rbd[@]}"
+                fi
+                find "${COMMAND_OUTPUT_DIR}" -name "rbd_vol_and_snap_info_*.part" -print0 | xargs -0 cat >>"${COMMAND_OUTPUT_FILE}"
+                find "${COMMAND_OUTPUT_DIR}" -name "rbd_vol_and_snap_info_*.part" -print0 | xargs -0 rm -f
+                
+                dbglogf "${CEPH_GATHER_DBGLOG}" "Collecting rbd trash ls for rados namespace: %s %s\n" "${bp}" "${rns}"
+                COMMAND_OUTPUT_FILE=${COMMAND_OUTPUT_DIR}/rbd_trash_ls_"${bp}"_"${rns}"
+                { timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rbd trash ls --pool $bp --namespace $rns --format=json" >>"${COMMAND_OUTPUT_FILE}"; } >>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-trash-ls-"${bp}"-"${rns}"-json-debug.log 2>&1 &
+                
+                COMMAND_OUTPUT_FILE=${COMMAND_OUTPUT_DIR}/rados_rbd_objects_"${bp}"_"${rns}"
+                # List omapkeys in csi.volumes.default in each block pool
+                pvcobjs=$(timeout 60 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "rados listomapkeys csi.volumes.default --pool=${bp} --namespace ${rns}")
+                # Get the omap details of each PVC object
+                for pvcobj in $pvcobjs; do
+                    { printf "Name of the pvc object: %s\n" "${pvcobj}" >>"${COMMAND_OUTPUT_FILE}"; }
```
@yati1998 is it what we have here is 100% duplication with additional `--namespace` flag. 
IMO if we add any command specific to rbd we need to add it in two different place which is maintenance burden and error prone.

jarrpa replied:
As I mentioned in [my previous comment](https://github.com/red-hat-storage/odf-must-gather/pull/150#issuecomment-2077081376), I don't think it's worth jumping through the hoops of having a subfunction, at least for this PR. I'm fine asking to clean that up in a follow-up PR, but as it stands this works and I'd like to move it along.

yati1998 replied:
yeah true that, Even I am currently working on it. 
Even if we add functions, we will have to repeat the loops and commands
The only cleanup possible is, adding commands to variables and using those variables at both the places.
Seems just an extra piece of work to me.

@jarrpa I am alreadyon that cleanup, will update the PR, if the collection works fine. orelse @Madhu-1 we can go ahead with this.

yati1998 commented:
@jarrpa @Madhu-1 I have addressed the comments to add helper functions and some more of cleanup. Please do review the PR. Also I have tested the changes and could collect all the required resources both in case of radosnamespace as well as in it's absence

Madhu-1 commented on collection-scripts/gather_ceph_resources:
```diff
@@ -199,77 +199,141 @@ for ns in $namespaces; do
 
-        # Collecting rados object information for RBD PVs and snapshots
-        COMMAND_OUTPUT_FILE=${COMMAND_OUTPUT_DIR}/rados_rbd_objects
+        collect_image_info(){
+            dbglogf "${CEPH_GATHER_DBGLOG}" "collecting vol and snapshot info for ${image}"
+            {   
+                if [ -n "${3}" ]; then
+                    image_info_p="$image_info $2 --pool $1 --namespace $3"
+                    status_info_p="$status_info $2 --pool $1 --namespace $3"
+                    snap_info_p="$snap_info $2 --pool $1 --namespace $3"
+                else
+                    image_info_p="$image_info $2 --pool $1"
+                    status_info_p="$status_info $2 --pool $1"
+                    snap_info_p="$snap_info $2 --pool $1"
+                fi
+                printf "Collecting image info for: %s/%s\n" "$1" "$2"
+                timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$image_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-"$2"-debug.log
+                printf "Collecting image status for: %s/%s\n" "$1" "$2"
+                timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$status_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-status-"$2"-debug.log
+                printf "Collecting snap info for: %s/%s\n" "$1" "$2"
+                timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$snap_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-snap-json-"$2"-debug.log
+            } >>"${COMMAND_OUTPUT_DIR}"/rbd_vol_and_snap_info_"$2".part &
+            pids_rbd+=($!)
+        }
```
cant we do something like below
```
collect_image_info(){
    dbglogf "${CEPH_GATHER_DBGLOG}" "collecting vol and snapshot info for ${image}"
    
    collect_info() {
        local pool=$1
        local image=$2
        local namespace=$3
        
        local ns_flag=""
        if [ -n "${namespace}" ]; then
            ns_flag="--namespace $namespace"
        fi

        local image_info_p="$image_info $image --pool $pool $ns_flag"
        local status_info_p="$status_info $image --pool $pool $ns_flag"
        local snap_info_p="$snap_info $image --pool $pool $ns_flag"

        printf "Collecting image info for: %s/%s\n" "$pool" "$image"
        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$image_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-"$image"-debug.log
        printf "Collecting image status for: %s/%s\n" "$pool" "$image"
        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$status_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-image-status-"$image"-debug.log
        printf "Collecting snap info for: %s/%s\n" "$pool" "$image"
        timeout 120 oc -n "${ns}" exec "${HOSTNAME}"-helper -- bash -c "$snap_info_p" 2>>"${COMMAND_ERR_OUTPUT_DIR}"/gather-rbd-snap-json-"$image"-debug.log
    }

    collect_info "$1" "$2" "$3" >>"${COMMAND_OUTPUT_DIR}"/rbd_vol_and_snap_info_"$2".part &
    pids_rbd+=($!)
}
```

Madhu-1 reviewed (commented):
the suggested changes LGTM but as am not an expert in this repo not able to confirm its covering all the required items.

yati1998 commented:
> the suggested changes LGTM but as am not an expert in this repo not able to confirm its covering all the required items.

Thanks @Madhu-1 for your review, it collects all the info that is collected for other images. In case you guys find anything missing further while debugging , feel free to open a bug/ issue, will add it.

yati1998 commented:
@black-dragon74 can you please review the PR

black-dragon74 commented on collection-scripts/gather_namespaced_resources:
```diff
@@ -20,7 +20,7 @@ resources=()
 # resources+=(storagesystem)
 
-# collect OB/OBC resoureces
+# collect OB/OBC resources
```
There is a PR open for fixing this CI failure at: #164 

It would make sense to exclude this change from this PR (as it is unrelated anyways)

yati1998 replied:
okay sure, since ci was failing, I added it.

black-dragon74 commented on collection-scripts/gather_ceph_resources:
```diff
@@ -199,77 +199,147 @@ for ns in $namespaces; do
+                    local status_info_p="$status_info $image --pool $pool $ns_flag"
+                    local snap_info_p="$snap_info $image --pool $pool $ns_flag"
+
+                    printf "Collecting image info for: %s/%s\n" "$pool" "$image"
```
Are we intentionally not using `dbglog`? Using dbglog will print to the screen and append to the file as well.

yati1998 replied:
yes, already printing the msg above, not required to print again, but just adding to file works, so I guess it's fine

black-dragon74 replied:
Edit: Please ignore the above comment.

black-dragon74 reviewed (changes requested):
Please fix the indentation for `collect_info` function at line: 205

We can get rid of unnecessary braces surrounding commands at lines mentioned below as they serve no purpose.

- 249
- 275
- 287
- 322
- 324
- 334
- 336

Regards

black-dragon74 commented:
Also, please mark the open conversations are resolved?

black-dragon74 commented:
> This seems like a superfluous cleanup that doesn't need to be part of this PR. I'd say it's best to leave them as-is and remove them in a follow-up PR as desired.

That is exactly what it is. Completely fine with taking it in a separate PR :)

Regards

yati1998 commented:
/cherry-pick release-4.15

yati1998 commented:
/cherry-pick release-4.16
